src/Core/Services/MetadataProviders/SqlMetadataProvider.cs

```diff
@@ -1666,52 +1666,98 @@ IEnumerable<List<ForeignKeyDefinition>> foreignKeysForAllTargetEntities
+        /// Loops over all the foreign key definitions defined for the target entity in the source entity's definition
+        /// and adds to the set of validated FK definitions:
+        /// 1. All the FK definitions which actually map to a foreign key constraint defined in the database.
+        /// In such a case, if the source/target fields are also provided in the config, they are given precedence over the FK constraint.
+        /// 2. FK definitions for custom relationships defined by the user in the configuration file where no FK constraint exists between
+        /// the pair of (source, target) entities.
+        /// </summary>
+        /// <param name="fKDefinitionsToTarget">List of FK definitions defined from source to target.</param>
+        /// <returns>List of validated FK definitions from source to target.</returns>
+        private List<ForeignKeyDefinition> GetValidatedFKs(
+            List<ForeignKeyDefinition> fKDefinitionsToTarget)
+        {
+            List<ForeignKeyDefinition> validatedFKDefinitionsToTarget = new();
+            foreach (ForeignKeyDefinition fKDefinitionToTarget in fKDefinitionsToTarget)
+            {
+                // This code block adds FK definitions between source and target entities when there is an FK constraint defined
+                // in the database, either from source->target or target->source entities or both.
 
-                                // Only add the referenced columns if they have not been
-                                // specified in the configuration file.
-                                if (fk.ReferencedColumns.Count == 0)
-                                {
-                                    fk.ReferencedColumns.AddRange(inferredDefinition.ReferencedColumns);
-                                }
-                            }
-                        }
+                // Add the referencing and referenced columns for this foreign key definition for the target.
+                if (PairToFkDefinition is not null &&
+                    PairToFkDefinition.TryGetValue(fKDefinitionToTarget.Pair, out ForeignKeyDefinition? inferredFKDefinition))
+                {
+                    // Being here indicates that we inferred an FK constraint for the current FK.
+                    // If the referencing and referenced columns count > 0,
+                    // we have already gathered this information from the runtime config.
+                    if (fKDefinitionToTarget.ReferencingColumns.Count > 0 && fKDefinitionToTarget.ReferencedColumns.Count > 0)
+                    {
+                        validatedFKDefinitionsToTarget.Add(fKDefinitionToTarget);
+                    }
+                    // Only add the referencing/referenced columns if they have not been
+                    // specified in the configuration file.
+                    else
+                    {
+                        validatedFKDefinitionsToTarget.Add(inferredFKDefinition);
```

Comment:
The current logic checks individually, for `ReferencingColumns.Count == 0` before adding `ReferencingColumn` information. Likewise, for `ReferencingColumns.Count`. 

AFAICT, this is possible only `source.fields` is empty when `target.fields` is non-empty (or vice-versa). Such a setup is invalid. It looks like it should be okay to populate the inferred details from the database without those checks. 
Do you see any side-effects of populating without those checks? @Aniruddh25

Reply:
Yeah I and Ani had a thorough discussion around this and we came to an agreement that this change makes logical sense and should be fine.

Reply:
Please add details around the same as comments or in your PR description

Reply:
- > All the FK definitions which actually map to a foreign key constraint defined in the database. In such a case, the source/target fields are also provided in the config, they are given precedence over the FK constraint.

I had added this detail initially. Works?

Reply:
The check for count of referencing/referenced columns being 0 is removed before adding `ReferencingColumn`/`ReferencedColumn` information. Adding details about why it is okay to remove this check will be helpful in the future if we were to look at this logic in the future.

Reply:
the concern raised in this thread should be resolved by #2139 to catch mismatch in number of config defined referencing/referenced columns (target/source fields).